allow torch._decomp.get_decompositions to include cpp decomps

bdhirsh opened the pull request:
This is more of an RFC - totally happy to scrap this PR for other alternatives.

Context: @andrewor14 has been working on getting the QAT workflow to use pre_dispatch tracing ([doc](https://docs.google.com/document/d/1esdgl7oT65haLa3QsNGMBrAd2RqiqZO6nF_oxrCCu-k/edit)). One issue he hit: QAT would like to get out a graph that is **pre-autograd**, but it also wants any CompositeImplicitAutograd ops to be **decomposed**.

This doesn't happen today: pre_dispatch tracing turns off CompositeImplicitAutograd ops by default, since the assumption is that you don't want to trace through **any** dispatcher code.

In theory this shouldn't be too big of a deal: `torch.export` lets you specify a decomp table, so if you want to run CompositeImplicit ops, you can just include them in the table. There isn't really an easy way to do this today, though.

Tentative fix: in this PR, I add an optional kwarg `include_cpp_decomps=False` to `torch._decomp.get_decompositions`, which will also return callables for any cpp CompositeImplicitAutograd ops, if any exist (and there isn't already a python decomp for that op). The callable is just a thin shim around `OpOverload.decompose()`.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105865
* #106329

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

Differential Revision: [D48456084](https://our.internmc.facebook.com/intern/diff/D48456084)

bdhirsh committed:
allow torch._decomp.get_decompositions to include cpp decomps

[ghstack-poisoned]

bdhirsh committed:
Update on "allow torch._decomp.get_decompositions to include cpp decomps"

This is more of an RFC - totally happy to scrap this PR for other alternatives.

Context: andrewor14 has been working on getting the QAT workflow to use pre_dispatch tracing ([doc](https://docs.google.com/document/d/1esdgl7oT65haLa3QsNGMBrAd2RqiqZO6nF_oxrCCu-k/edit)). One issue he hit: QAT would like to get out a graph that is **pre-autograd**, but it also wants any CompositeImplicitAutograd ops to be **decomposed**.

This doesn't happen today: pre_dispatch tracing turns off CompositeImplicitAutograd ops by default, since the assumption is that you don't want to trace through **any** dispatcher code.

In theory this shouldn't be too big of a deal: `torch.export` lets you specify a decomp table, so if you want to run CompositeImplicit ops, you can just include them in the table. There isn't really an easy way to do this today, though.

Tentative fix: in this PR, I add an optional kwarg `include_cpp_decomps=False` to `torch._decomp.get_decompositions`, which will also return callables for any cpp CompositeImplicitAutograd ops, if any exist (and there isn't already a python decomp for that op). The callable is just a thin shim around `OpOverload.decompose()`.




[ghstack-poisoned]

ezyang reviewed (approved):
Perhaps a leeetle test?

fduwjj commented:
@fduwjj has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47760074).

bdhirsh committed:
Update on "allow torch._decomp.get_decompositions to include cpp decomps"

This is more of an RFC - totally happy to scrap this PR for other alternatives.

Context: andrewor14 has been working on getting the QAT workflow to use pre_dispatch tracing ([doc](https://docs.google.com/document/d/1esdgl7oT65haLa3QsNGMBrAd2RqiqZO6nF_oxrCCu-k/edit)). One issue he hit: QAT would like to get out a graph that is **pre-autograd**, but it also wants any CompositeImplicitAutograd ops to be **decomposed**.

This doesn't happen today: pre_dispatch tracing turns off CompositeImplicitAutograd ops by default, since the assumption is that you don't want to trace through **any** dispatcher code.

In theory this shouldn't be too big of a deal: `torch.export` lets you specify a decomp table, so if you want to run CompositeImplicit ops, you can just include them in the table. There isn't really an easy way to do this today, though.

Tentative fix: in this PR, I add an optional kwarg `include_cpp_decomps=False` to `torch._decomp.get_decompositions`, which will also return callables for any cpp CompositeImplicitAutograd ops, if any exist (and there isn't already a python decomp for that op). The callable is just a thin shim around `OpOverload.decompose()`.


Differential Revision: [D47760074](https://our.internmc.facebook.com/intern/diff/D47760074)

[ghstack-poisoned]

bdhirsh committed:
Update on "allow torch._decomp.get_decompositions to include cpp decomps"

This is more of an RFC - totally happy to scrap this PR for other alternatives.

Context: andrewor14 has been working on getting the QAT workflow to use pre_dispatch tracing ([doc](https://docs.google.com/document/d/1esdgl7oT65haLa3QsNGMBrAd2RqiqZO6nF_oxrCCu-k/edit)). One issue he hit: QAT would like to get out a graph that is **pre-autograd**, but it also wants any CompositeImplicitAutograd ops to be **decomposed**.

This doesn't happen today: pre_dispatch tracing turns off CompositeImplicitAutograd ops by default, since the assumption is that you don't want to trace through **any** dispatcher code.

In theory this shouldn't be too big of a deal: `torch.export` lets you specify a decomp table, so if you want to run CompositeImplicit ops, you can just include them in the table. There isn't really an easy way to do this today, though.

Tentative fix: in this PR, I add an optional kwarg `include_cpp_decomps=False` to `torch._decomp.get_decompositions`, which will also return callables for any cpp CompositeImplicitAutograd ops, if any exist (and there isn't already a python decomp for that op). The callable is just a thin shim around `OpOverload.decompose()`.


Differential Revision: [D47760074](https://our.internmc.facebook.com/intern/diff/D47760074)

[ghstack-poisoned]

bdhirsh committed:
Update on "allow torch._decomp.get_decompositions to include cpp decomps"

This is more of an RFC - totally happy to scrap this PR for other alternatives.

Context: andrewor14 has been working on getting the QAT workflow to use pre_dispatch tracing ([doc](https://docs.google.com/document/d/1esdgl7oT65haLa3QsNGMBrAd2RqiqZO6nF_oxrCCu-k/edit)). One issue he hit: QAT would like to get out a graph that is **pre-autograd**, but it also wants any CompositeImplicitAutograd ops to be **decomposed**.

This doesn't happen today: pre_dispatch tracing turns off CompositeImplicitAutograd ops by default, since the assumption is that you don't want to trace through **any** dispatcher code.

In theory this shouldn't be too big of a deal: `torch.export` lets you specify a decomp table, so if you want to run CompositeImplicit ops, you can just include them in the table. There isn't really an easy way to do this today, though.

Tentative fix: in this PR, I add an optional kwarg `include_cpp_decomps=False` to `torch._decomp.get_decompositions`, which will also return callables for any cpp CompositeImplicitAutograd ops, if any exist (and there isn't already a python decomp for that op). The callable is just a thin shim around `OpOverload.decompose()`.


Differential Revision: [D47760074](https://our.internmc.facebook.com/intern/diff/D47760074)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

bdhirsh commented on torch/_decomp/__init__.py:
```diff
@@ -139,6 +139,8 @@ def register(op):
     aten_ops: Sequence[Union[OpOverload, OpOverloadPacket]],
     type: str = "post_autograd",
+    *,
+    prefer_cpp_composites: bool = False,
```
cc @ezyang I ended up... adding this flag, lmk what you think. This issue is that @andrewor14 would like a way to set up his decomp table for `pre_dispatch`, such that the **exact same** set of decomps end up running as if you had run vanilla `make_fx()`.

When you run vanilla `make_fx()`, C++ CompositeImplicit decomps get priority over any python decomps, so this flag basically lets you explicitly ask for the same behavior.

ezyang replied:
okey dokey

bdhirsh committed:
Update on "allow torch._decomp.get_decompositions to include cpp decomps"

This is more of an RFC - totally happy to scrap this PR for other alternatives.

Context: andrewor14 has been working on getting the QAT workflow to use pre_dispatch tracing ([doc](https://docs.google.com/document/d/1esdgl7oT65haLa3QsNGMBrAd2RqiqZO6nF_oxrCCu-k/edit)). One issue he hit: QAT would like to get out a graph that is **pre-autograd**, but it also wants any CompositeImplicitAutograd ops to be **decomposed**.

This doesn't happen today: pre_dispatch tracing turns off CompositeImplicitAutograd ops by default, since the assumption is that you don't want to trace through **any** dispatcher code.

In theory this shouldn't be too big of a deal: `torch.export` lets you specify a decomp table, so if you want to run CompositeImplicit ops, you can just include them in the table. There isn't really an easy way to do this today, though.

Tentative fix: in this PR, I add an optional kwarg `include_cpp_decomps=False` to `torch._decomp.get_decompositions`, which will also return callables for any cpp CompositeImplicitAutograd ops, if any exist (and there isn't already a python decomp for that op). The callable is just a thin shim around `OpOverload.decompose()`.


Differential Revision: [D47760074](https://our.internmc.facebook.com/intern/diff/D47760074)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

BowenBao commented:
iiuc this allows selective decomposing CompositeImplicitAutograd ops by using `pre_dispatch` and a filtered decomp table. Is it possible to run subsequent functionalization on the graph without decomposing CompositeImplicitAutograd ops?

bdhirsh commented:
Hey @BowenBao - late reply, but we recently started talking about (eventually) exposing a pre_dispatch, functionalized IR in the longer term. It'll probably be a few months before its ready, but it will be functionalized, and not have to decompose all CompositeImplicitAutograd ops (only the subset that are "problematic", like `aten::reshape`, `aten::native_batch_norm`, and a few others (these ops "sometimes alias" and "sometimes mutate", which is not something we can functionalize)

andrewor14 committed:
Update on "allow torch._decomp.get_decompositions to include cpp decomps"

This is more of an RFC - totally happy to scrap this PR for other alternatives.

Context: andrewor14 has been working on getting the QAT workflow to use pre_dispatch tracing ([doc](https://docs.google.com/document/d/1esdgl7oT65haLa3QsNGMBrAd2RqiqZO6nF_oxrCCu-k/edit)). One issue he hit: QAT would like to get out a graph that is **pre-autograd**, but it also wants any CompositeImplicitAutograd ops to be **decomposed**.

This doesn't happen today: pre_dispatch tracing turns off CompositeImplicitAutograd ops by default, since the assumption is that you don't want to trace through **any** dispatcher code.

In theory this shouldn't be too big of a deal: `torch.export` lets you specify a decomp table, so if you want to run CompositeImplicit ops, you can just include them in the table. There isn't really an easy way to do this today, though.

Tentative fix: in this PR, I add an optional kwarg `include_cpp_decomps=False` to `torch._decomp.get_decompositions`, which will also return callables for any cpp CompositeImplicitAutograd ops, if any exist (and there isn't already a python decomp for that op). The callable is just a thin shim around `OpOverload.decompose()`.


Differential Revision: [D47760074](https://our.internmc.facebook.com/intern/diff/D47760074)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

andrewor14 committed:
Update on "allow torch._decomp.get_decompositions to include cpp decomps"

This is more of an RFC - totally happy to scrap this PR for other alternatives.

Context: andrewor14 has been working on getting the QAT workflow to use pre_dispatch tracing ([doc](https://docs.google.com/document/d/1esdgl7oT65haLa3QsNGMBrAd2RqiqZO6nF_oxrCCu-k/edit)). One issue he hit: QAT would like to get out a graph that is **pre-autograd**, but it also wants any CompositeImplicitAutograd ops to be **decomposed**.

This doesn't happen today: pre_dispatch tracing turns off CompositeImplicitAutograd ops by default, since the assumption is that you don't want to trace through **any** dispatcher code.

In theory this shouldn't be too big of a deal: `torch.export` lets you specify a decomp table, so if you want to run CompositeImplicit ops, you can just include them in the table. There isn't really an easy way to do this today, though.

Tentative fix: in this PR, I add an optional kwarg `include_cpp_decomps=False` to `torch._decomp.get_decompositions`, which will also return callables for any cpp CompositeImplicitAutograd ops, if any exist (and there isn't already a python decomp for that op). The callable is just a thin shim around `OpOverload.decompose()`.


Differential Revision: [D47760074](https://our.internmc.facebook.com/intern/diff/D47760074)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

andrewor14 commented on torch/_inductor/lowering.py:
```diff
@@ -1692,10 +1709,8 @@ def apply_constraint(arg, fx_arg):
-make_fallback(aten.convolution_backward, constrain_to_fx_strides)
 make_fallback(aten._cudnn_rnn, require_dense)
 make_fallback(aten._cudnn_rnn_backward, require_contiguous)
-make_fallback(aten.cumsum, require_dense, warn=False)
```
cc @eellison just wanted to check if this is OK. It's needed to get the tests to pass

eellison replied:
Is there a way to look at the cpp decompositions ?

What does `aten.cumprod`, and `aten.sort` get mapped to now ? 

We need to replicate the existing `constrain_to_fx_strides` for convolution_backward but maybe that's happening organically and it's appearing as   as `aten.convolution`?

bdhirsh replied:
@eellison these "new" decomps are all just coming from the original C++ CompositeImplicitAutograd decomps from the dispatcher. I can try to find a few of them so we can check.

Alternatively: what do you think of updating the fallback logic to "allow" a fallback, if we detect that the op in question has a CompositeImplicitAutograd registration? (so we'd be saying "actual no, inductor shouldn't just use CompositeImplicitAutograd decomps over fallbacks by default")

bdhirsh replied:
> Alternatively: what do you think of updating the fallback logic to "allow" a fallback, if we detect that the op in question has a CompositeImplicitAutograd registration? (so we'd be saying "actual no, inductor shouldn't just use CompositeImplicitAutograd decomps over fallbacks by default")

This feels like less work though, compared to auditing the fallbacks in inductor and replacing the overload packet fallbacks with more finegrained fallbacks for overloads. So if this sounds reasonable to you, I'll probably just do this

andrewor14 committed:
Update on "allow torch._decomp.get_decompositions to include cpp decomps"

This is more of an RFC - totally happy to scrap this PR for other alternatives.

Context: andrewor14 has been working on getting the QAT workflow to use pre_dispatch tracing ([doc](https://docs.google.com/document/d/1esdgl7oT65haLa3QsNGMBrAd2RqiqZO6nF_oxrCCu-k/edit)). One issue he hit: QAT would like to get out a graph that is **pre-autograd**, but it also wants any CompositeImplicitAutograd ops to be **decomposed**.

This doesn't happen today: pre_dispatch tracing turns off CompositeImplicitAutograd ops by default, since the assumption is that you don't want to trace through **any** dispatcher code.

In theory this shouldn't be too big of a deal: `torch.export` lets you specify a decomp table, so if you want to run CompositeImplicit ops, you can just include them in the table. There isn't really an easy way to do this today, though.

Tentative fix: in this PR, I add an optional kwarg `include_cpp_decomps=False` to `torch._decomp.get_decompositions`, which will also return callables for any cpp CompositeImplicitAutograd ops, if any exist (and there isn't already a python decomp for that op). The callable is just a thin shim around `OpOverload.decompose()`.


Differential Revision: [D47760074](https://our.internmc.facebook.com/intern/diff/D47760074)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

andrewor14 committed:
Update on "allow torch._decomp.get_decompositions to include cpp decomps"

This is more of an RFC - totally happy to scrap this PR for other alternatives.

Context: andrewor14 has been working on getting the QAT workflow to use pre_dispatch tracing ([doc](https://docs.google.com/document/d/1esdgl7oT65haLa3QsNGMBrAd2RqiqZO6nF_oxrCCu-k/edit)). One issue he hit: QAT would like to get out a graph that is **pre-autograd**, but it also wants any CompositeImplicitAutograd ops to be **decomposed**.

This doesn't happen today: pre_dispatch tracing turns off CompositeImplicitAutograd ops by default, since the assumption is that you don't want to trace through **any** dispatcher code.

In theory this shouldn't be too big of a deal: `torch.export` lets you specify a decomp table, so if you want to run CompositeImplicit ops, you can just include them in the table. There isn't really an easy way to do this today, though.

Tentative fix: in this PR, I add an optional kwarg `include_cpp_decomps=False` to `torch._decomp.get_decompositions`, which will also return callables for any cpp CompositeImplicitAutograd ops, if any exist (and there isn't already a python decomp for that op). The callable is just a thin shim around `OpOverload.decompose()`.


Differential Revision: [D47760074](https://our.internmc.facebook.com/intern/diff/D47760074)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

andrewor14 committed:
Update on "allow torch._decomp.get_decompositions to include cpp decomps"

This is more of an RFC - totally happy to scrap this PR for other alternatives.

Context: andrewor14 has been working on getting the QAT workflow to use pre_dispatch tracing ([doc](https://docs.google.com/document/d/1esdgl7oT65haLa3QsNGMBrAd2RqiqZO6nF_oxrCCu-k/edit)). One issue he hit: QAT would like to get out a graph that is **pre-autograd**, but it also wants any CompositeImplicitAutograd ops to be **decomposed**.

This doesn't happen today: pre_dispatch tracing turns off CompositeImplicitAutograd ops by default, since the assumption is that you don't want to trace through **any** dispatcher code.

In theory this shouldn't be too big of a deal: `torch.export` lets you specify a decomp table, so if you want to run CompositeImplicit ops, you can just include them in the table. There isn't really an easy way to do this today, though.

Tentative fix: in this PR, I add an optional kwarg `include_cpp_decomps=False` to `torch._decomp.get_decompositions`, which will also return callables for any cpp CompositeImplicitAutograd ops, if any exist (and there isn't already a python decomp for that op). The callable is just a thin shim around `OpOverload.decompose()`.


Differential Revision: [D47760074](https://our.internmc.facebook.com/intern/diff/D47760074)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48456084).

andrewor14 committed:
Update on "allow torch._decomp.get_decompositions to include cpp decomps"

This is more of an RFC - totally happy to scrap this PR for other alternatives.

Context: andrewor14 has been working on getting the QAT workflow to use pre_dispatch tracing ([doc](https://docs.google.com/document/d/1esdgl7oT65haLa3QsNGMBrAd2RqiqZO6nF_oxrCCu-k/edit)). One issue he hit: QAT would like to get out a graph that is **pre-autograd**, but it also wants any CompositeImplicitAutograd ops to be **decomposed**.

This doesn't happen today: pre_dispatch tracing turns off CompositeImplicitAutograd ops by default, since the assumption is that you don't want to trace through **any** dispatcher code.

In theory this shouldn't be too big of a deal: `torch.export` lets you specify a decomp table, so if you want to run CompositeImplicit ops, you can just include them in the table. There isn't really an easy way to do this today, though.

Tentative fix: in this PR, I add an optional kwarg `include_cpp_decomps=False` to `torch._decomp.get_decompositions`, which will also return callables for any cpp CompositeImplicitAutograd ops, if any exist (and there isn't already a python decomp for that op). The callable is just a thin shim around `OpOverload.decompose()`.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D48456084](https://our.internmc.facebook.com/intern/diff/D48456084)

[ghstack-poisoned]

andrewor14 committed:
Update on "allow torch._decomp.get_decompositions to include cpp decomps"

This is more of an RFC - totally happy to scrap this PR for other alternatives.

Context: andrewor14 has been working on getting the QAT workflow to use pre_dispatch tracing ([doc](https://docs.google.com/document/d/1esdgl7oT65haLa3QsNGMBrAd2RqiqZO6nF_oxrCCu-k/edit)). One issue he hit: QAT would like to get out a graph that is **pre-autograd**, but it also wants any CompositeImplicitAutograd ops to be **decomposed**.

This doesn't happen today: pre_dispatch tracing turns off CompositeImplicitAutograd ops by default, since the assumption is that you don't want to trace through **any** dispatcher code.

In theory this shouldn't be too big of a deal: `torch.export` lets you specify a decomp table, so if you want to run CompositeImplicit ops, you can just include them in the table. There isn't really an easy way to do this today, though.

Tentative fix: in this PR, I add an optional kwarg `include_cpp_decomps=False` to `torch._decomp.get_decompositions`, which will also return callables for any cpp CompositeImplicitAutograd ops, if any exist (and there isn't already a python decomp for that op). The callable is just a thin shim around `OpOverload.decompose()`.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D48456084](https://our.internmc.facebook.com/intern/diff/D48456084)

[ghstack-poisoned]

andrewor14 committed:
Update on "allow torch._decomp.get_decompositions to include cpp decomps"

This is more of an RFC - totally happy to scrap this PR for other alternatives.

Context: andrewor14 has been working on getting the QAT workflow to use pre_dispatch tracing ([doc](https://docs.google.com/document/d/1esdgl7oT65haLa3QsNGMBrAd2RqiqZO6nF_oxrCCu-k/edit)). One issue he hit: QAT would like to get out a graph that is **pre-autograd**, but it also wants any CompositeImplicitAutograd ops to be **decomposed**.

This doesn't happen today: pre_dispatch tracing turns off CompositeImplicitAutograd ops by default, since the assumption is that you don't want to trace through **any** dispatcher code.

In theory this shouldn't be too big of a deal: `torch.export` lets you specify a decomp table, so if you want to run CompositeImplicit ops, you can just include them in the table. There isn't really an easy way to do this today, though.

Tentative fix: in this PR, I add an optional kwarg `include_cpp_decomps=False` to `torch._decomp.get_decompositions`, which will also return callables for any cpp CompositeImplicitAutograd ops, if any exist (and there isn't already a python decomp for that op). The callable is just a thin shim around `OpOverload.decompose()`.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D48456084](https://our.internmc.facebook.com/intern/diff/D48456084)

[ghstack-poisoned]

andrewor14 committed:
Update on "allow torch._decomp.get_decompositions to include cpp decomps"

This is more of an RFC - totally happy to scrap this PR for other alternatives.

Context: andrewor14 has been working on getting the QAT workflow to use pre_dispatch tracing ([doc](https://docs.google.com/document/d/1esdgl7oT65haLa3QsNGMBrAd2RqiqZO6nF_oxrCCu-k/edit)). One issue he hit: QAT would like to get out a graph that is **pre-autograd**, but it also wants any CompositeImplicitAutograd ops to be **decomposed**.

This doesn't happen today: pre_dispatch tracing turns off CompositeImplicitAutograd ops by default, since the assumption is that you don't want to trace through **any** dispatcher code.

In theory this shouldn't be too big of a deal: `torch.export` lets you specify a decomp table, so if you want to run CompositeImplicit ops, you can just include them in the table. There isn't really an easy way to do this today, though.

Tentative fix: in this PR, I add an optional kwarg `include_cpp_decomps=False` to `torch._decomp.get_decompositions`, which will also return callables for any cpp CompositeImplicitAutograd ops, if any exist (and there isn't already a python decomp for that op). The callable is just a thin shim around `OpOverload.decompose()`.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D48456084](https://our.internmc.facebook.com/intern/diff/D48456084)

[ghstack-poisoned]

andrewor14 committed:
Update on "allow torch._decomp.get_decompositions to include cpp decomps"

This is more of an RFC - totally happy to scrap this PR for other alternatives.

Context: andrewor14 has been working on getting the QAT workflow to use pre_dispatch tracing ([doc](https://docs.google.com/document/d/1esdgl7oT65haLa3QsNGMBrAd2RqiqZO6nF_oxrCCu-k/edit)). One issue he hit: QAT would like to get out a graph that is **pre-autograd**, but it also wants any CompositeImplicitAutograd ops to be **decomposed**.

This doesn't happen today: pre_dispatch tracing turns off CompositeImplicitAutograd ops by default, since the assumption is that you don't want to trace through **any** dispatcher code.

In theory this shouldn't be too big of a deal: `torch.export` lets you specify a decomp table, so if you want to run CompositeImplicit ops, you can just include them in the table. There isn't really an easy way to do this today, though.

Tentative fix: in this PR, I add an optional kwarg `include_cpp_decomps=False` to `torch._decomp.get_decompositions`, which will also return callables for any cpp CompositeImplicitAutograd ops, if any exist (and there isn't already a python decomp for that op). The callable is just a thin shim around `OpOverload.decompose()`.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D48456084](https://our.internmc.facebook.com/intern/diff/D48456084)

[ghstack-poisoned]

andrewor14 committed:
Update on "allow torch._decomp.get_decompositions to include cpp decomps"

This is more of an RFC - totally happy to scrap this PR for other alternatives.

Context: andrewor14 has been working on getting the QAT workflow to use pre_dispatch tracing ([doc](https://docs.google.com/document/d/1esdgl7oT65haLa3QsNGMBrAd2RqiqZO6nF_oxrCCu-k/edit)). One issue he hit: QAT would like to get out a graph that is **pre-autograd**, but it also wants any CompositeImplicitAutograd ops to be **decomposed**.

This doesn't happen today: pre_dispatch tracing turns off CompositeImplicitAutograd ops by default, since the assumption is that you don't want to trace through **any** dispatcher code.

In theory this shouldn't be too big of a deal: `torch.export` lets you specify a decomp table, so if you want to run CompositeImplicit ops, you can just include them in the table. There isn't really an easy way to do this today, though.

Tentative fix: in this PR, I add an optional kwarg `include_cpp_decomps=False` to `torch._decomp.get_decompositions`, which will also return callables for any cpp CompositeImplicitAutograd ops, if any exist (and there isn't already a python decomp for that op). The callable is just a thin shim around `OpOverload.decompose()`.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D48456084](https://our.internmc.facebook.com/intern/diff/D48456084)

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/lowering.py:
```diff
@@ -1382,6 +1383,22 @@ def check_skip_condition(node, is_output):
+    if isinstance(kernel, torch._ops.OpOverloadPacket):
+        for overload_name in kernel._overload_names:
+            overload = getattr(kernel, overload_name)
+            make_fallback(overload)
```
```suggestion
            make_fallback(overload, layout_constraint=layout_constraint, warn=warn)
```

andrewor14 committed:
Update on "allow torch._decomp.get_decompositions to include cpp decomps"

This is more of an RFC - totally happy to scrap this PR for other alternatives.

Context: andrewor14 has been working on getting the QAT workflow to use pre_dispatch tracing ([doc](https://docs.google.com/document/d/1esdgl7oT65haLa3QsNGMBrAd2RqiqZO6nF_oxrCCu-k/edit)). One issue he hit: QAT would like to get out a graph that is **pre-autograd**, but it also wants any CompositeImplicitAutograd ops to be **decomposed**.

This doesn't happen today: pre_dispatch tracing turns off CompositeImplicitAutograd ops by default, since the assumption is that you don't want to trace through **any** dispatcher code.

In theory this shouldn't be too big of a deal: `torch.export` lets you specify a decomp table, so if you want to run CompositeImplicit ops, you can just include them in the table. There isn't really an easy way to do this today, though.

Tentative fix: in this PR, I add an optional kwarg `include_cpp_decomps=False` to `torch._decomp.get_decompositions`, which will also return callables for any cpp CompositeImplicitAutograd ops, if any exist (and there isn't already a python decomp for that op). The callable is just a thin shim around `OpOverload.decompose()`.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D48456084](https://our.internmc.facebook.com/intern/diff/D48456084)

[ghstack-poisoned]

bdhirsh committed:
Update on "allow torch._decomp.get_decompositions to include cpp decomps"

This is more of an RFC - totally happy to scrap this PR for other alternatives.

Context: andrewor14 has been working on getting the QAT workflow to use pre_dispatch tracing ([doc](https://docs.google.com/document/d/1esdgl7oT65haLa3QsNGMBrAd2RqiqZO6nF_oxrCCu-k/edit)). One issue he hit: QAT would like to get out a graph that is **pre-autograd**, but it also wants any CompositeImplicitAutograd ops to be **decomposed**.

This doesn't happen today: pre_dispatch tracing turns off CompositeImplicitAutograd ops by default, since the assumption is that you don't want to trace through **any** dispatcher code.

In theory this shouldn't be too big of a deal: `torch.export` lets you specify a decomp table, so if you want to run CompositeImplicit ops, you can just include them in the table. There isn't really an easy way to do this today, though.

Tentative fix: in this PR, I add an optional kwarg `include_cpp_decomps=False` to `torch._decomp.get_decompositions`, which will also return callables for any cpp CompositeImplicitAutograd ops, if any exist (and there isn't already a python decomp for that op). The callable is just a thin shim around `OpOverload.decompose()`.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D48456084](https://our.internmc.facebook.com/intern/diff/D48456084)

[ghstack-poisoned]

bdhirsh committed:
Update on "allow torch._decomp.get_decompositions to include cpp decomps"

This is more of an RFC - totally happy to scrap this PR for other alternatives.

Context: andrewor14 has been working on getting the QAT workflow to use pre_dispatch tracing ([doc](https://docs.google.com/document/d/1esdgl7oT65haLa3QsNGMBrAd2RqiqZO6nF_oxrCCu-k/edit)). One issue he hit: QAT would like to get out a graph that is **pre-autograd**, but it also wants any CompositeImplicitAutograd ops to be **decomposed**.

This doesn't happen today: pre_dispatch tracing turns off CompositeImplicitAutograd ops by default, since the assumption is that you don't want to trace through **any** dispatcher code.

In theory this shouldn't be too big of a deal: `torch.export` lets you specify a decomp table, so if you want to run CompositeImplicit ops, you can just include them in the table. There isn't really an easy way to do this today, though.

Tentative fix: in this PR, I add an optional kwarg `include_cpp_decomps=False` to `torch._decomp.get_decompositions`, which will also return callables for any cpp CompositeImplicitAutograd ops, if any exist (and there isn't already a python decomp for that op). The callable is just a thin shim around `OpOverload.decompose()`.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D48456084](https://our.internmc.facebook.com/intern/diff/D48456084)

[ghstack-poisoned]

bdhirsh commented:
Agreed with @andrewor14 that QAT no longer needs this now that pre-dispatch is not decomposing most CompositeImplicitAutograd ops, and can use this patch instead:
```
import torch

decomps = {
    torch.ops.aten.batch_norm.default: lambda *args: torch.ops.aten.batch_norm.default.decompose(*args),
    torch.ops.aten._batch_norm_impl_index.default: lambda *args: torch.ops.aten._batch_norm_impl_index.default.decompose(*args),
    torch.ops.aten.native_batch_norm.default: lambda *args: torch.ops.aten.native_batch_norm.default.decompose(*args),
}

m = torch.nn.BatchNorm2d(4)
m.eval()
x = torch.randn(4, 4, 4, 4)
fx_g = torch._dynamo.export(m, x, aten_graph=True, pre_dispatch=True, decomposition_table=decomps)
print(fx_g.graph_module)
```
